### ISSUE TITLE
Generate PDF pages for JetForm repeater field entries

### ISSUE THREAD

**Issue by ghost** [CrocoBlock/suggestions#7727]
The feature I am looking for is regarding JetForm and the generate PDF functionality. It works great when creating a template using normal data fields, but it is limited when it comes to repeater fields.

Currently when selecting a repeater field in a template it prints the values like this:
![image](https://github.com/user-attachments/assets/f13091bf-924e-4a1e-8397-9a9a82dbeba3)

So all repeater fields just unterneath each other.

And this brings me to the problem - when for example offering tours or hotels, in the booking process the person booking it often has to fill out a form where the names of all travellers are added into a repeater field. Like on the left side of the previous screenshot or like in this screenshot from your own page: 
![image](https://github.com/user-attachments/assets/85c2abf2-4c4a-4c6c-933d-7d36a454b360)

Now for such tours, tickets or hotel vouchers are quite common. **So the feature that is needed is to be able to generate a PDF with an own page for each entry of the repeater.** For example someone books a tour for 3 persons, fills in all 3 names into the repeater, I would then like to send out a PDF with 3 pages (tickets), with each page holding the name of one person (and some other data next to it which is the same on each page).

[Example PDF.pdf](https://github.com/user-attachments/files/16902347/Example.PDF.pdf)
You can look at my attached PDF to see what I mean. On the first page there is all the normal booking information, the name of the person who booked the tour on the top and the 3 persons the tour was booke for at the bottom - this would be possible with your current tool, but then there are 3 separate pages for each person. Hope this makes it clear what I think is needed as a feature.

So in general what is needed, is a better and more flexible handling of repeater fields when it comes to generating PDFs.

As a way to implement this I thought of something like a new block like "repeater page" where I type in the name of the repeater field and then within this block I can create an own template for one page with the values from the repeater. And when the PDF gets generated it adds the amount of pages with that template for the amount of repeater entries to the PDF.

Other possible implementation I thought of could be to create a normal second template where I type in the field names of the repeater. And then in the actual template a add this "repeater page" block and type in the repeater field name and tell the block to use the second template for each of the repeater entries.

Hope my ideas where clear as well as my explanation of this missing feature, which I think would be great to be added.

The addition of generating PDFs was huge and I appreciate your possibilities when it comes to tour or hotel bookings. Especially the option of making a repeater field for multiple travellers, but sadly repeater fields can not be handled quite well in PDFs at the moment so a feature like this would be a great addition and could offer great possibilties like generating personalized hotel vouchers, travel information or tickets for each person.

**Comment by buithucdong** [comment]
Yeah, I have the same problems when try to make a pdf to print bill with repeat field 
![Nh_p th_ng tin __n h_ng _ Qu_n l_ __n h_ng](https://github.com/user-attachments/assets/b3dc7d6f-ddba-445a-95e7-02ac9f391929)
![Don-hang-tu-Form-6_pdf](https://github.com/user-attachments/assets/628005ad-1b78-4946-938e-bacf2ebfc514)

**Comment by zakwan-1** [comment]
I have same issue! Does it solved?

**Comment by vinnyCalixto** [comment]
same issue! anyone solve it?

**Comment by moxet** [comment]
You have to insert the following code in your functions.php or phpsnippet

```
/*
for example:
%repeater_content(repeater1|br)%
{item_count}: %text1%
%repeater_content%
or
%repeater_content(repeater1|nl)%
or
%repeater_content(repeater1|, )%
etc.
for the first line
you can also use it in Hidden fields (turn Render in HTML off), putting that in Manual input with the prefix mailmacros::
*/

class JFB_Readable_Repeater {
	
	public function __construct() {
		add_filter( 'jet-form-builder/rich-content', array( $this, 'convert' ), 0 );
		add_action( 'jet-form-builder/form-handler/before-send', array( $this, 'do_macros_in_fields' ) );
	}
	
	public function is_new() {
		return function_exists( 'jet_fb_context' );
	}
	
	public function get_form_data() {
		
		if ( $this->is_new() ) {
			$form_data = jet_fb_context()->resolve_request();	
		} else {
			$form_data = jet_fb_action_handler()->request_data;
		}
		
		return $form_data;
		
	}
	
	public function do_macros_in_fields( $form_handler ) {
		
		$form_data = $this->get_form_data();

		foreach ( $form_data as $key => $value ) {
			
			if ( ! is_scalar( $value ) || false === strpos( $value, 'mailmacros::' ) ) {
				continue;
			}
			
			$value = str_replace( 'mailmacros::', '', $value );
		
			if ( $this->is_new() ) {
				jet_fb_context()->update_request( $this->convert( $value ), $key );
			} else {
				$action_handler->request_data[ $key ] = $this->convert( $value );
			}
			
		}

	}
	
	public function convert( $content ) {
	
		$pattern = '/%repeater_content\((.+?)\)%(.+?)%repeater_content%/ms';

		$form_data = $this->get_form_data();
		
		if ( class_exists( '\Jet_Form_Builder\Classes\Macros_Parser' ) ) {
			$parser = new \Jet_Form_Builder\Classes\Macros_Parser();	
		} else {
			$parser = new \JFB_Modules\Rich_Content\Macros_Parser();
		}
		
		$content = preg_replace_callback( $pattern, function( $matches ) use ( $form_data, $parser ) {

			$params = explode( '|', $matches[1] );

			$repeater_key = $params[0];
			$delimiter    = $params[1] ?? ', ';

			$delimiter = str_replace( "br", "<br>", $delimiter );
			$delimiter = str_replace( "nl", "\n", $delimiter );

			$repeater_value = $form_data[ $repeater_key ];

			if ( empty( $repeater_value ) || ! is_array( $repeater_value ) ) {
				return '';
			}

			$template = $matches[2];

			$repeater_content = array();

			$count = 1;

			foreach ( $repeater_value as $item ) {
				$item_content = str_replace( '{item_count}', $count, $template );
				$repeater_content[] = trim( $parser->parse_macros( $item_content, $item ), "\n\r" );
				$count++;
			}

			return implode( $delimiter, $repeater_content );

		}, $content );

		return $content;

	}
	
}

new JFB_Readable_Repeater();
```
